articles/112_generated_interfaces_cpp.md

```diff
@@ -0,0 +1,132 @@
+The filenames use lowercase alphanumeric characters with underscores for separating words and end with either `.hpp` or `.cpp`.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> describe separate header files
+</div>
+
+
+### Messages
+
+A message results in a templated `struct` with the same name followed by an underscore.
+The single template argument is the allocator for the data structure.
+
+For ease of use there is a `typedef` with the same name as the message which uses a default allocator (e.g. `std::allocator`).
+
+
+#### Members
+
+The struct has same-named member variables for every field of the message.
+For each field a `typedef` is created which is named after the member with a leading underscore and a trailing `_type`.
+
+
+#### Constants
+
+Numeric constants are defined as `enums` within the struct.
+
+All other constants are declared as `static const` members in the struct and their values are defined outside of the struct.
+
+
+#### Getters
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> const and non-const getters
+</div>
+
+
+#### Setters
+
+Beside the direct access to the member variables for each member a *setter* method is generated.
+They are named after the members with a leading `set_`.
+The setter methods have a single argument to pass the value for the member variable.
+Each setter methods returns the struct itself to enable (method chaining)[https://isocpp.org/wiki/faq/ctors#named-parameter-idiom}.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> Need non-colliding naming
+</div>
+
+
+#### Constructors
+
+The *default constructor* initializes all members with their default value.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> Should we use in-class member initializers for this instead of the ctor initializer syntax?
+</div>
+
+Optionally the constructor can be invoked with an allocator.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> The *copy constructor* initializes all members based on a reference instance. `=default`?
+</div>
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> *Move constructor*? `=default`?
+</div>
+
+The struct has no constructor with positional arguments for the members.
+The short reason for this is that if code would reply on positional arguments to construct data objects changing a message definition would break existing code in various ways.
+Since this would discourage to evolve message definitions the data structures should be populared by setting the members separately, e.g. using the setter methods.
+
+
+#### Operators
+
+The comparison operators `==` and `!=` perform the comparison on a per member basis.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> *Copy / move assignment operators*? `=default`?
+</div>
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> *Swap*? `=default`?
```

Comment:
Is there a default swap? I thought this only existed in vectors and similar std containers.